### PR TITLE
AM-52: reset password policy settings

### PR DESCRIPTION
**Pull request by mstasgravitee**
AM-52

### CODE REVIEW

File: gravitee-am-service/src/main/java/io/gravitee/am/service/model/PatchDomain.java [hunk 1]
```diff
@@ -220,8 +221,14 @@ public Domain patch(Domain _toPatch) {
...
-        if (this.passwordSettings != null) {
-            this.passwordSettings.ifPresent(ps -> toPatch.setPasswordSettings(ps.patch(toPatch.getPasswordSettings())));
+        if (passwordSettings != null && passwordSettings.isPresent()) {
+            var ps = passwordSettings.get();
+            if (ps.getInherited() != null && ps.getInherited().isPresent() && Boolean.TRUE.equals(ps.getInherited().get())) {
+                toPatch.setPasswordSettings(null);
+                setPasswordSettings(Optional.empty());
+            } else {
+                toPatch.setPasswordSettings(ps.patch(toPatch.getPasswordSettings()));
+            }
```
Comment (leleueri) [inline gravitee-am-service/src/main/java/io/gravitee/am/service/model/PatchDomain.java:231]: **todo**: Not sure about this way to manage the reset. You are considering the presence of the inherited boolean but everywhere else into the code for patch/update action we consider the Optional.isEmpty() as the condition to reset the value to null. 

Basically, in the request payload:
* if the field is missing, the Java bean will have a null for this attribute,  in that case we do not change the value
* if the field is set to null, the Java bean will have a Optional.empty() for this attribute,  in that case we set the value to null
* if the field is not null, the Java bean will have a Optional.from() for this attribute,  in that case we change the value with the content of the Optional.

The change should be something like this IMO to be consistent

```suggestion
        if (passwordSettings != null) {
            if (passwordSettings.isEmpty()) {
                toPatch.setPasswordSettings(null);
            } else {
                toPatch.setPasswordSettings(this.passwordSettings.get().patch(toPatch.getPasswordSettings()));
            }
```
wdyt ?

File: gravitee-am-service/src/main/java/io/gravitee/am/service/model/PatchDomain.java [hunk 2]
```diff
@@ -260,7 +267,7 @@ public Set<Permission> getRequiredPermissions() {
             requiredPermissions.add(Permission.DOMAIN_SETTINGS);
         }
 
-        if(alertEnabled != null && alertEnabled.isPresent()) {
+        if (alertEnabled != null && alertEnabled.isPresent()) {
```
Comment (leleueri) [inline gravitee-am-service/src/main/java/io/gravitee/am/service/model/PatchDomain.java:270]: **question**: are we sure that this change will not break the current behaviour ? :thinking: This change will probably allow to disable the Alert feature een if the user doesn't have the DOMAIN_ALERT permission. I think this change should be revert. (I can see that we test the `isPresent` in the previous statement but don't know if we have to apply the same for alerts)

File: gravitee-am-ui/src/app/domain/settings/password-policy/domain-password-policy.component.ts [hunk 3]
```diff
@@ -132,12 +133,38 @@ export class DomainPasswordPolicyComponent implements OnInit {
...
+    this.dialogService
+      .confirm('Reset password policy', 'Are you sure you want to reset your password policy settings and use the default one?')
+      .subscribe((res) => {
+        if (res) {
+          this.reset();
+        }
+      });
+  }
+  reset(): void {
+    this.domainService.patchPasswordSettings(this.domainId, { passwordSettings: { inherited: true } }).subscribe(() => {
```
Comment (leleueri) [inline gravitee-am-ui/src/app/domain/settings/password-policy/domain-password-policy.component.ts:155]: **todo**: based on the comment on gravitee-am-service/src/main/java/io/gravitee/am/service/model/PatchDomain.java we should set passwordSettings to null.

```suggestion
    this.domainService.patchPasswordSettings(this.domainId, { passwordSettings: null }).subscribe(() => {
```

**Review by leleueri — CHANGES_REQUESTED** [review]

File: gravitee-am-service/src/main/java/io/gravitee/am/service/model/PatchDomain.java (hunk 2 above)
Comment (mstasgravitee) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/model/PatchDomain.java:270]: It has been added only a space to fix format here.

Comment (leleueri) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/model/PatchDomain.java:270]: :tired_face: sorry about that one

File: gravitee-am-service/src/main/java/io/gravitee/am/service/model/PatchDomain.java (hunk 1 above)
Comment (mstasgravitee) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/model/PatchDomain.java:231]: image
image

Comment (leleueri) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/model/PatchDomain.java:231]: ok, as discussed we can rely on null.

**Pull request merged**